src/dmd/root/array.d

```diff
@@ -591,3 +591,68 @@ unittest
     static assert (!is(typeof(Array!(OtherStruct).init.sort())));
     static assert (!is(typeof(Array!(OtherStruct).init.sort!pred)));
 }
+
+/**
+ * Filters the given array based on the given predicate.
+ *
+ * Returns: a new array containing only elements for which the predicate returns
+ *  `true`
+ */
+inout(T)[] filter(alias predicate, T)(inout T[] array)
+if (is(ReturnType!(typeof((T t) => predicate(t))) == bool))
+{
+    inout(T)[] result;
```

Comment:
Shouldn't something like `Appender!(T[]) result;` be used here instead and then return `result.data` at the end of the function?

Reply:
`Appender` is part of Phobos and is not allowed in the compiler proper.

Reply:
That's gonna create lots of garbage. I suggest we copy Appender into dmd.

Reply:
It's not going to create anymore garbage then it's already doing.